### PR TITLE
Ensure that MeasureInvalidated properly propagates up

### LINKED ISSUE #25650: [Bug] iOS: CollectionView items don't change their size when item content size is changed
### Description

https://github.com/dotnet/maui/issues/3643 regressed between net8 and net9

### PR DESCRIPTION
**Pull request by PureWeen**
### Description of Change

The MeasureInvalidated call is propagating up from Label/Button/Editor too early. 
It needs to propagate after the handler has finished processing the changes

### Issues Fixed

Fixes #25650

### CODE REVIEW

File: src/Controls/src/Core/Label/Label.cs [hunk 1]
```diff
@@ -382,6 +382,17 @@ void ITextElement.OnTextColorPropertyChanged(Color oldValue, Color newValue)
 		void ITextElement.OnCharacterSpacingPropertyChanged(double oldValue, double newValue) =>
 			InvalidateMeasureIfLabelSizeable();
 
+		TextDecorations ILabel.TextDecorations
+		{ 
+			get
+			{
+				if (Text is not null)
```
Comment (albyrock87) [inline src/Controls/src/Core/Label/Label.cs:389]: Is this condition here for some not evident reason?

Comment (PureWeen) [reply src/Controls/src/Core/Label/Label.cs:389]: Whoops :-/ fixed

**Comment by albyrock87** [comment]
Reminder for me to revert this or part of this in https://github.com/dotnet/maui/pull/25664

File: src/Controls/src/Core/Label/Label.Android.cs [hunk 2]
```diff
@@ -39,6 +39,7 @@ private protected override void OnHandlerChangedCore()
 		public static void MapText(ILabelHandler handler, Label label)
 		{
 			Platform.TextViewExtensions.UpdateText(handler.PlatformView, label);
+			InvalidateMeasureIfNecessary(label);
 		}
```
Comment (jsuarezruiz) [inline src/Controls/src/Core/Label/Label.Android.cs:43]: Do we know why is now required to invalidate measure here? https://github.com/dotnet/maui/blob/bd614274484c7fa0fd7fac008c551932613b75b3/src/Controls/src/Core/Label/Label.cs#L422 Is InvalidateMeasureIfLabelSizeable, invoked with any modification in the Label bounds (text, character spacing, etc) working?

File: src/Controls/src/Core/Button/Button.iOS.cs [hunk 3]
```diff
@@ -464,6 +464,7 @@ private static void MapPadding(IButtonHandler handler, Button button)
 		public static void MapText(IButtonHandler handler, Button button)
 		{
 			handler.PlatformView?.UpdateText(button);
+			button.FireMeasureChanged(Internals.InvalidationTrigger.MeasureChanged, 0);
```
Comment (jsuarezruiz) [inline src/Controls/src/Core/Button/Button.iOS.cs:467]: Like happens with Editor or Label, is not necessary to invalidate measure on other platforms too?

File: src/Controls/tests/TestCases.HostApp/Issues/Issue3643.xaml.cs [hunk 4]
```diff
@@ -0,0 +1,100 @@
+﻿using System.Collections.ObjectModel;
+using System.ComponentModel;
+
+namespace Maui.Controls.Sample.Issues
+{
+
+	[Issue(IssueTracker.Github, 3643, "CollectionView items don't change their size when item content size is changed", PlatformAffected.iOS)]
+	public partial class Issue3643 : ContentPage
```
Comment (jsuarezruiz) [inline src/Controls/tests/TestCases.HostApp/Issues/Issue3643.xaml.cs:8]: Nice tests, not in this scope of this PR, but would be a good idea to create another test based on this one, doing the same tests but with all the Views.

**Comment by jsuarezruiz** [comment]
Tested Issue3643 on Android and iOS, and works as expected with the changes.

File: src/Controls/src/Core/Button/Button.iOS.cs (hunk 3 above)
Comment (PureWeen) [reply src/Controls/src/Core/Button/Button.iOS.cs:467]: Yea, I was trying to be minimal but actually I just noticed that we already have this on the button BPs

image

So it was already firing via that. 

I deleted the code from the BP and moved this so all the mappers now just fire.

Good find!

File: src/Controls/src/Core/Label/Label.Android.cs (hunk 2 above)
Comment (PureWeen) [reply src/Controls/src/Core/Label/Label.Android.cs:43]: The initial problem here is that the order of events was happening like this

BP fires property changed
InvalidateMeasure Fires too early for CV to catch and remeasure
Mapperr Runs (this is what changes the text)

If you look at Label.cs you'll see where i've basically just moved the invalidation call so now it fires in this order

BP fires property changed
Mapperr Runs (this is what changes the text)
InvalidateMeasure Fires now with the text changed

File: src/Controls/tests/TestCases.HostApp/Issues/Issue3643.xaml.cs (hunk 4 above)
Comment (PureWeen) [reply src/Controls/tests/TestCases.HostApp/Issues/Issue3643.xaml.cs:8]: Yea, I'd like to look through all the CV tests that we're porting over to see if there's a set that cover this

**Review by jsuarezruiz — DISMISSED** [review]

**Review by jsuarezruiz — COMMENTED** [review]
Not sure (yet) if related, are failing some Layouts tests:

```
 Failed UnconstrainedStarRowWithMultipleStarColumnsAllowsTextToGrow [< 1 ms]
  Error Message:
   a: 87 b: 174 tolerance: 2
  Stack Trace:
     at Microsoft.Maui.Controls.Core.UnitTests.GridTests.AssertEqualWithTolerance(Double a, Double b, Double tolerance) in /_/src/Compatibility/Core/tests/Compatibility.UnitTests/GridTests.cs:line 448
   at Microsoft.Maui.Controls.Core.UnitTests.GridTests.UnconstrainedStarRowWithMultipleStarColumnsAllowsTextToGrow() in /_/src/Compatibility/Core/tests/Compatibility.UnitTests/GridTests.cs:line 439
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
```

https://github.com/dotnet/maui/blob/bd614274484c7fa0fd7fac008c551932613b75b3/src/Compatibility/Core/tests/Compatibility.UnitTests/GridTests.cs#L393

**Comment by PureWeen** [comment]
Pausing on this for now

I was thinking this scenario was isolated to label and button, but we generally always propagate InvalidationMeasure before calling the mappers and I don't quite understand why. 

it looks like we've been trickling this fix of forcing the mappers to fire from the BP property changed https://github.com/dotnet/maui/pull/6665 in order to fix timing issues. 

And this PR is just another ring in the chain instead of fixing the actual problem. 

Or I don't understand why we fire invalidatemeasure before mappers run

Thanks @albyrock87 !!

**Pull request closed without merging**